File: php/127-word-ladder.php
<?php
class Solution {

    /**
     * @param String $beginWord
     * @param String $endWord
     * @param String[] $wordList
     * @return Integer
     */
    function ladderLength($beginWord, $endWord, $wordList) {
        $visited = [];
        $dict = [];
        $queue = [[$beginWord, 1]];

        // init dict
        foreach ($wordList as $w) {
            $len = strlen($w);
            for ($i = 0; $i < $len; $i++) {
                $dkey = substr($w, 0, $i) . '*' . substr($w, $i+1);
                $dict[$dkey][] = $w;
            }
        }
        //BF
        while ($queue) {
            [$word, $step] = array_shift($queue);
            //replace one character
            $len = strlen($word);
            for ($i = 0; $i < $len; $i++) {
                $dkey = substr($word, 0, $i) . '*' . substr($word, $i+1);
                if (!isset($dict[$dkey])) {
                    continue;
                }
                foreach ($dict[$dkey] as $w) {
                    if ($w == $endWord) {
                        return $step + 1;
                    }
                    if (!isset($visited[$w])) {
                        array_push($queue, [$w, $step+1]);
                        $visited[$w] = 1;
                    }
                }
            }
        }
        return 0;
    }
}
$begin = 'nanny';
$end = 'aloud';
$list = [];
echo (new Solution())->ladderLength($begin, $end, $list);

File: go/127-word-ladder.go
package main

type QueueWord struct {
	word string
	step int
}

func ladderLength(beginWord string, endWord string, wordList []string) int {
	visited := make(map[string]bool)
	dict := make(map[string][]string)
	var queue []QueueWord
	queue = append(queue, QueueWord{beginWord, 1})
	// init dict
	for _, word := range wordList {
		for i, _ := range word {
			key := word[0:i] + "_" + word[i+1:]
			dict[key] = append(dict[key], word)
		}
	}
	for len(queue) > 0 {
		w := queue[0]
		queue = queue[1:]
		word := w.word
		step := w.step
		for i, _ := range word {
			key := word[0:i] + "_" + word[i+1:]
			for _, dictword := range dict[key] {
				if endWord == dictword {
					return step + 1
				}
				if _, ok := visited[dictword]; !ok {
					queue = append(queue, QueueWord{dictword, step + 1})
					visited[dictword] = true
				}
			}
		}
	}
	return 0
}
